### PR TITLE
request.go update

### PR DESCRIPTION
**Pull request by shazbert**
## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

Fixes 422 issue with Poloniex when out of order requests are sent. This ensures correct fifo insertion into worker channel buffer for concurrent requests. All exchanges updated accordingly.

# How Has This Been Tested?

# AUTH NEEDS TO BE TESTED ON ALL EXCHANGES BEFORE MERGE

basic go test ./... used locally

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool 
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Travis with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/request/request.go [hunk 1]
```diff
@@ -391,29 +396,39 @@ func (r *Requester) worker() {
 }
 
 // SendPayload handles sending HTTP/HTTPS requests
-func (r *Requester) SendPayload(method, path string, headers map[string]string, body io.Reader, result interface{}, authRequest, verbose bool) error {
+func (r *Requester) SendPayload(method, path string, headers map[string]string, body io.Reader, result interface{}, authRequest, nonceEnabled, verbose bool) error {
+	if !nonceEnabled {
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/request/request.go]: It's locked when you derive the nonce. That lock goes across functions up until passed into the work buffer. Have you found it locks down when running?

[the remark this answers is not in the document]
Comment (gloriousCode) [reply exchanges/request/request.go]: Oh sweet, I misread and thought your unlocked after generating the nonce, that solves my issues and concerns

File: exchanges/coinut/coinut.go [hunk 2]
```diff
@@ -338,20 +338,17 @@ func (c *COINUT) SendHTTPRequest(apiRequest string, params map[string]interface{
...
 	if params == nil {
 		params = map[string]interface{}{}
 	}
-	params["nonce"] = c.Nonce.Get()
+	params["nonce"] = n
 	params["request"] = apiRequest
 
 	payload, err := common.JSONEncode(params)
 	if err != nil {
+		c.Requester.UnlockFifo()
```
Comment (thrasher-) [inline exchanges/coinut/coinut.go]: This looks like it should be removed

File: exchanges/poloniex/poloniex_test.go [hunk 3]
```diff
@@ -19,67 +20,78 @@ const (
...
-	poloniexConfig, err := cfg.GetExchangeConfig("Poloniex")
-	if err != nil {
-		t.Error("Test Failed - Poloniex Setup() init error")
-	}
-	poloniexConfig.AuthenticatedAPISupport = true
-	poloniexConfig.APIKey = apiKey
-	poloniexConfig.APISecret = apiSecret
-
-	p.Setup(&poloniexConfig)
+	mtx.Lock()
```
Comment (thrasher-) [inline exchanges/poloniex/poloniex_test.go]: the sync.Mutex code can be be rm -rf'd

File: exchanges/request/request.go [hunk 4]
```diff
@@ -455,6 +471,34 @@ func (r *Requester) SendPayload(method, path string, headers map[string]string,
 	return resp.Error
 }
 
+// GetNonce returns a nonce for requests. This locks and enforces concurrent
+// nonce FIFO on the buffered job channel
+func (r *Requester) GetNonce(isNano bool) int64 {
```
Comment (thrasher-) [inline exchanges/request/request.go]: What do you think about returning a nonce.Value instead, that way you can use the built in String() method so you don't have to worry about strconv.

Comment (shazbert) [reply exchanges/request/request.go]: Good suggestion will change.

**Pull request merged**